plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/N4JSHover.java

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (c) 2017 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ui.editor;
+
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.jface.text.IRegion;
+import org.eclipse.jface.text.ITextViewer;
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.service.OperationCanceledError;
+import org.eclipse.xtext.ui.editor.hover.DispatchingEObjectTextHover;
+import org.eclipse.xtext.ui.editor.model.IXtextDocument;
+import org.eclipse.xtext.ui.editor.model.XtextDocumentUtil;
+import org.eclipse.xtext.util.Pair;
+import org.eclipse.xtext.util.concurrent.IUnitOfWork;
+
+/**
+ * Minor customization of the default {@link DispatchingEObjectTextHover} to avoid blocking the UI thread.
+ */
+public class N4JSHover extends DispatchingEObjectTextHover {
+
+	/**
+	 * Method copied from super class with only a minor change: call to "readOnly" changed to "tryReadOnly".
+	 */
+	@Override
+	public IRegion getHoverRegion(final ITextViewer textViewer, final int offset) {
+		IXtextDocument xtextDocument = XtextDocumentUtil.get(textViewer);
+		if (xtextDocument == null)
+			return null;
+		// TODO this is being called on change in the UI-thread. Not a good idea to do such expensive stuff.
+		// returning the region on a per token basis would be better.
+		try {
+			return ((N4JSDocument) xtextDocument).tryReadOnly(new IUnitOfWork<IRegion, XtextResource>() {
```

Comment:
Should we use org.eclipse.xtext.util.concurrent.CancelableUnitOfWork<R, P> ?

Reply:
as discussed